### DISCUSSION TITLE
Fine-Tuning Model with Covariates and Target Time Series

### DISCUSSION THREAD

**Discussion by thisthq** [SalesforceAIResearch/uni2ts#53] [Q&A]
Thank you for your contributions and those of your colleagues.

I am working on a multivariate time series forecasting task. My data consists of 31 time series, including 30 time series as covariates and 1 time series as a target that I need to forecast. Can I fine-tune MOIRAI with these 30 covariates and 1 target as described above?

Thank you very much.

**Comment by gorold** [comment]
Sure, you could do that. You would need to process your data into the appropriate format.

**Reply by Riofd** [reply]
> Sure, you could do that. You would need to process your data into the appropriate format.

Hello, I noticed that there is no annotation of covariates in the example of  [dataset processing](https://github.com/SalesforceAIResearch/uni2ts/blob/main/example/prepare_data.ipynb). Does it need to specify the covariate column names during the dataset processing? and need any other settings to make minor adjustments?

**Reply by Houyikai** [reply]
```
from typing import Generator, Any, Dict
import pandas as pd
from datasets import Features, Value, Sequence

def example_gen_func() -> Generator[Dict[str, Any], None, None]:
    for i in range(len(df_after.columns)-1):
        yield {
            "target": df_after.iloc[:, i].to_numpy(),  # array of shape (time,)
            "start": df_after.index[0],
            "freq": pd.infer_freq(df_after.index),
            "item_id": f"item_{i}",
            "past_feat_dynamic_real": df_after.iloc[:, -1].to_numpy(),
        }

features = Features({
    "target": Sequence(Value("float32")),
    "start": Value("timestamp[s]"),
    "freq": Value("string"),
    "item_id": Value("string"),
    "past_feat_dynamic_real": Sequence(Value("int32")),
})

```

Is this code correct? The last column is the covariate

**Reply by Houyikai** [reply]
> > Sure, you could do that. You would need to process your data into the appropriate format.
> 
> Hello, I noticed that there is no annotation of covariates in the example of [dataset processing](https://github.com/SalesforceAIResearch/uni2ts/blob/main/example/prepare_data.ipynb). Does it need to specify the covariate column names during the dataset processing? and need any other settings to make minor adjustments?

did you solve it？

**Reply by Houyikai** [reply]
What I really want to use is ”feat_dynamic_real“, but when I read the training code, I found that this part of processing is not available at all. My understanding is that as long as MaskedPrediction does not handle feat_dynamic_real field, attention can have all the context of it, this feature can be implemented, but it is not included in the code.

**Reply by Riofd** [reply]
I used `feat_dynamic_real` in `example_gen_func()`, and it successfully generated my own datasets. I use it for fine-tune, and it shows improvement with the help of covariates.

**Reply by Houyikai** [reply]
> I used `feat_dynamic_real` in `example_gen_func()`, and it successfully generated my own datasets. I use it for fine-tune, and it shows improvement with the help of covariates.

Thank you. I tried as well, but it didn't show any improvements, so I thought I had done it wrong.

**Reply by Riofd** [reply]
Did you successfully generate dataset with correct `feat_dynamic_real`? If so, I believe you can fine-tune the model with it. Moreover, my fine-tuned improvement is compared with not fine-tuning that performs very bad on my dataset.

**Reply by Houyikai** [reply]
我测是不行。我读他代码，没发现处理feat_dynamic_real这个字段，只用了target。我感觉这个代码有问题，指定是哪不对。

**Reply by gorold** [reply]
The data generation script looks correct. For the current pre-training/fine-tuning implementation, feat_dynamic_real are not included, but with some minor edits, they should be supported.

**Reply by lilyniu** [reply]
I would like to proceed with including feat_dynamic_real in the training process. Could you provide some guidance on how we can make the necessary modifications to support this feature? Specifically, what kind of minor edits do you think would be required to accommodate feat_dynamic_real?

**Reply by gorold** [reply]
you'll need to process the data appropriately, and update the transforms to consider feat_dynamic_real too.

**Reply by liyujia011025** [reply]
> Did you successfully generate dataset with correct ? If so, I believe you can fine-tune the model with it. Moreover, my fine-tuned improvement is compared with not fine-tuning that performs very bad on my dataset.`feat_dynamic_real`

Hello expert:
May I ask if you made the fine-tuning using the process and code in the official README file? It's strange that after fine-tuning with my own data according to this operation, the effect is even worse than not fine-tuning at all. Did you make any specific changes to hyperparameter settings or other content during the fine-tuning process? Also, may I ask if you only fine tuned certain layers or all layers during the fine-tuning process?
Looking forward to your reply, thank you very much!

**Reply by liyujia011025** [reply]
> > 我在 中使用了 ，它成功地生成了我自己的数据集。我用它来微调，它在协变量的帮助下显示出改进。`feat_dynamic_real``example_gen_func()`
> 
> 谢谢。我也试了一下，但没有显示出任何改善，所以我认为我做错了。

Hello expert:
May I ask if your fine-tuning has been effective now? Did you make the fine-tuning using the process and code in the official README file? It's strange that after fine-tuning with my own data according to this operation, the effect is even worse than not fine-tuning at all. Did you make any specific changes to hyperparameter settings or other content during the fine-tuning process? Also, may I ask if you only fine tuned certain layers or all layers during the fine-tuning process?
Looking forward to your reply, thank you very much!

**Reply by liyujia011025** [reply]
> I would like to proceed with including feat_dynamic_real in the training process. Could you provide some guidance on how we can make the necessary modifications to support this feature? Specifically, what kind of minor edits do you think would be required to accommodate feat_dynamic_real?

![Uploading image.png…]()

**Comment by liyujia011025** [comment]
> Thank you for your contributions and those of your colleagues.
> 
> I am working on a multivariate time series forecasting task. My data consists of 31 time series, including 30 time series as covariates and 1 time series as a target that I need to forecast. Can I fine-tune MOIRAI with these 30 covariates and 1 target as described above?
> 
> Thank you very much.

Hello expert:
May I ask if your fine-tuning has been effective now? Did you make the fine-tuning using the process and code in the official README file? It's strange that after fine-tuning with my own data according to this operation, the effect is even worse than not fine-tuning at all. Did you make any specific changes to hyperparameter settings or other content during the fine-tuning process? Also, may I ask if you only fine tuned certain layers or all layers during the fine-tuning process?
Looking forward to your reply, thank you very much!
